﻿﻿﻿﻿﻿﻿﻿<h1  align="center">SimPlot++</h1>
<h2  align="center">Recombination detection software</h2>

<details open>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About the Project</a>
    </li>
    <li>
      <a href="#Installation">Installation</a>
      <ul>
        <li><a href="#Windows version">Windows version</a></li>
        <li><a href="#Linux/UNIX and Mac OS versions">Linux/UNIX and Mac OS versions</a></li>
      </ul>
    </li>
    <li> Available analyses</li>
      <ul>
        <li><a href="#Group Page">Group Page</a></li>
        <li><a href="#SimPlot analysis">SimPlot analysis</a></li>
        <li><a href="#Similarity network">Similarity network</a></li>
        <li><a href="#Bootscan analysis">Bootscan analysis</a></li>
        <li><a href="#Findsite">Findsite</a></li>
        <li><a href="#Detection of recombination">Detection of recombination</a></li>
      </ul>
  </ol>
</details>



# About The Project

Simplot++ is an open-source multi-platform application designed by Stéphane Samson, Étienne Lord and Vladimir Makarenkov (Université du Québec à Montréal). It is implemented in Python. SimPlot++ produces publication-ready SimPlot and BootScan plots using 43 nucleotide and 20 amino acid distance models. Intergenic and intragenic recombination events can be identified using Phi, χ2, NSS and Proportion tests, while diagnostic of recombination events can be visualized using interactive sequence similarity networks. SimPlot++ supports multicore processing and provides additional distance calculability diagnostics.

SimPlot++ offers the following features:

- Create and save consensus groups 
- Execute a SimPlot (sequence similarity plot) analysis
- Execute a BootScan analysis  
- Identify informative sites  
- Create and analyze sequence similarity networks
- Run Phi, NSS and χ2 statistical tests to detect recombination




# Installation
## Windows version

SimPlot++ for Windows is available either as an executable file or as a Python script.

### Executable
The Windows installer can be found in the [release](https://github.com/Stephane-S/Simplot_PlusPlus/releases). Click on the most recent `SimPlot++-x.x-win64.msi` file to download it.

### Python script
A `requirements.txt` file containing all required libraries is available in the GitHub repository.

Assuming Python 3.8 or higher is installed on the machine, the script should run well with the libraries installed.

<u>Here is an example of how to run the script in Windows:</u>
1. After downloading the source code, go to the folder containing `main.py`.
1. Create a new virtual environment (venv) in your Windows PowerShell using `python3 -m venv SimPlot++_venv `.
2. Still in the PowerShell, enter the new venv using `SimPlot++_env/Scripts/Activate.ps1`.
3. Install the required libraries using `pip install -r requirements.txt`.
4. Launch SimPlot++ using `python3 main.py`.

## Linux/UNIX and Mac OS versions
SimPlot++ is available as a Python script.

### Python script
A `requirements.txt` file containing all required libraries is available in the GitHub repository.

Assuming Python 3.8 or higher is installed on the machine, the script should run well with the libraries installed.

<u>Here is an example of how to run the script in Windows:</u>
1. After downloading the source code, go to the folder containing `main.py`.
2. If you do not have `virtualenv` installed, run `python3 -m pip install --user virtualenv`
3. Create a new virtual environment (venv) in your terminal using `python3 -m venv SimPlot++_env`.
4. Still in the terminal, enter the new venv using `source SimPlot++_env/bin/activate`.
5. Install the required libraries using `python3 -m pip install -r requirements.txt`.
6. Launch SimPlot++ using `python3 main.py`.



# getting started

## Group page
The group page allows users to load multiple sequences files (fasta, nexus, pir, phylip, stockholm or clustal format) and manually organize the individual sequences in groups. For each of the groups created by the user, a consensus sequence will be generated by SimPlot++. The % threshold for these consensus sequences can be modified in the preference tab. 
These consensus groups are essential for the SimPlot, BootScan and Similarity Network analysis. The groups can be saved to a nexus file to avoid recreating them every time.

### A typical workflow would look like this:

![group_page_step](https://github.com/Stephane-S/Simplot_PlusPlus/blob/master/Docs/images/group_page_steps.png)

1. Select your sequence file (DNA or AA) in an accepted format (Fasta, Nexus, Pir, Phylip, Stockholm or Clustal) through the **File browser** button
2. Once the file is loaded, the sequences ID will appear in the **Ungrouped Sequences** section, in the bottom right.
3. Groups can be created and deleted, and sequences ID can be moved from the **Ungrouped Sequences** section to the **Group section**
    - Groups can be renamed by clicking on it's given name.
    - Groups colors can be modified by clicking on it's colored circle.
    - A group that has no sequences will block the user from accessing most analyses available.
    - A minimum of two groups containing at least 1 sequence each are necessary to access every feature.
4. Once your groups are created, customized and are ready to use, it is suggested to save them locally using the **Save groups to .Nexus** button.
    - This feature will save your initial data in the nexus format along with the groups compositions, allowing you to skip the group creation process the next time you start SimPlot++
5. Once these steps are completed, you can execute your analysis of choice by selection it on the top bar.

Below is a summary of the steps presented:

![group page gif](https://github.com/Stephane-S/Simplot_PlusPlus/blob/master/Docs/images/group_page_steps_gif.gif)



# SimPlot analysis

## Methodology
A SimPlot analysis employs a window of a specified size and step to slide over the Multiple Sequences Alignment (MSA). Every subsequence covered by the window is extracted and a distance matrix based on a specified model is generated. These distance matrices are then used to plot the similarity of every sequence in the MSA against a specific sequence chosen by the user. The variations in similarity between the reference and the other sequences can help detect potential recombination events.

### Main features
- 43 DNA distance and 20 amino acids models models are available (includes models from [Biopython](https://biopython.org/) and [Cogent3](https://cogent3.org/)).
- A multiprocessing feature
- Matplotlib-based plots with a toolbar to easily customize and save the outputs in multiple formats.
- Plots can be viewed in a pop-up window (with the toolbar)
- A new quality control window will open an interactive HTML page to access additional information on the output quality.

### Settings
**Window length**: Determines the size of the sliding window 

**Step**: Determines how much the window will slide every iteration (with overlap)

**Strip gap**: Determines the maximum amount of ambiguous positions allowed in a consensus sequence to be included (optional)

**Multiprocessing**: Allows multiple windows to be analyzed simultaneously. Recommended for big data sets.

**Distance Model**: 43 DNA models and 20 protein models available

### Simplot analysis example

![Simplot analysis](https://github.com/Stephane-S/Simplot_PlusPlus/blob/master/Docs/images/simplot_gif.gif)

## Quality control tool
A new feature of SimPlot++ is the quality control feature for the Simplot analysis where the user can view gap-related data (such as the regions where the amount of gaps in the consensus sequence was higher than the gap threshold permitted). 

Overall data completeness (whether genetic distances were successfully computed or not in each window) can also be viewed through this feature.

Since certain models can throw errors when computing genetically distant sequences (divisions by zero, log of negative numbers, etc.), it is recommended to use this feature to check if such issues happened. If it did, it is recommended to modify the analysis parameters, consensus groups, consensus threshold, or the distance model used depending on the analysis. As a general rule, simpler models tend to be more lenient. 



# Similarity network

The similarity network analysis is an interactive representation of a SimPlot analysis where every group (without a chosen reference group) is represented by a node. These nodes are connected by their global (overall sequence) and local (per window) similarities. 

By adjusting the minimum similarity threshold required to show each of these edge types, it is possible to get a better overview look of the relationships between every group. Furthermore, the graph representation can be limited to a specific range of the full sequences, in order to isolate a gene or region of interest. 

Datatables allows the user to view the raw data represented visually in the graph. The results of the Proportion test are also available in a datatable (discussed in more details in the statistical methods section.

The graph data and visualization are stored in an HTML file that can be downloaded and shared while retaining their interactive features.

### Similarity network output example

![similarity network](https://github.com/Stephane-S/Simplot_PlusPlus/blob/master/Docs/images/newtwork_gif.gif)



# Bootscan analysis

## Methodology
Bootscanning is a pipeline consisting of 4 main steps, all done using a sliding window (just like the SimPlot analysis).
### Steps
1. The subsequences extracted from the consensus groups are bootstrapped N times.
2. For each of the N bootstrapped Multiple Sequences Alignments (MSA), a distance matrix is generated.
3. A phylogenetic tree is inferred for each distance matrix (either with Neighbor-Joining or UPGMA).
4. The conflicting phylogenetic signals are quantified and expressed as the % of trees where each sequence is the nearest neighbor of the reference sequence. 

### Main features
- 43 DNA models are available for generating the distance matrices
- A multiprocessing feature
- Matplotlib-based plots with a toolbar to easily customize and save the outputs in multiple formats
- Plots can be viewed in a pop-up window (with the toolbar)

### Settings
**Bootstrap**: Number of replicates to be generated for each dataset (window)

**Tree model**: Neighbor-Joining or UPGMA

**Window length**: Size of the sliding window 

**Step**: Sliding window advancement step

**Distance Model**: 43 DNA substitution models are available

**Multiprocessing**: Allows multiple windows to be analyzed simultaneously (recommended for large datasets).

### Bootscan analysis example

![Bootscan gif](https://github.com/Stephane-S/Simplot_PlusPlus/blob/master/Docs/images/bootscan_gif.gif)



# Findsite

The FindSite scan is used for locating possible regions of recombination by identifying informative sites. The first step of the analysis is to select a sequence suspected of having originated of a recombination event as well as two sequences of interest (one from each of the two suspected parental evolutionary lines) and a fourth from an evolutionarily unrelated sequence to the previous three (an outgroup). Informative sites will be identified as those where, at the same position, two of the sequences share the same nucleotide, and that the other two sequences share another different nucleotide.

### FindSite example
![Findsite](https://github.com/Stephane-S/Simplot_PlusPlus/blob/master/Docs/images/findsite.gif)



# Detection of recombination

Statistical tests for the detection of recombination events from the PhiPack package by Trevor Bruen have been implemented in SimPlot++. This implementation includes a graphical interface in order to simplify it's usage. The multiprocessing option is recommended for faster results. 

 The Phi , Phi profile, Max χ2 and NSS tests are available for both the ungrouped (raw sequences) and grouped consensus sequences. 

Additional information on these tests can be found [here](https://www.researchgate.net/publication/229005330_PhiPack_PHI_test_and_other_tests_of_recombination).



Additionally, a new proportion test has been designed as a complement to the traditional SimPlot analysis in order to identify the most likely mosaic regions in the grouped sequences. It is an indirect test based on the proportion of genetic distances extracted from the SimPlot distance matrices. The score is an indicator of the signal strength but should not be taken as a recombination signal. It's purpose is to help guide the user and highlight the most promising leads.

### Example Output
![recombination](https://github.com/Stephane-S/Simplot_PlusPlus/blob/master/Docs/images/recomb_gif.gif)



